pydantic/_internal/_core_utils.py

```diff
@@ -446,72 +447,52 @@ def count_refs(s: core_schema.CoreSchema, recurse: Recurse) -> core_schema.CoreS
-            # on its second encounter, we can safely stop the walk here.
             if current_recursion_ref_count[ref] != 0:
-                involved_in_recursion[ref] = True
+                involved_in_recursion.add(ref)
             return s
-
         current_recursion_ref_count[ref] += 1
         if 'serialization' in s:
-            # Even though this is a `'definition-ref'` schema, there might
-            # be more references inside the serialization schema:
             recurse(s, count_refs)
-
         next_s = definitions[ref]
-        visited: set[str] = set()
-        while next_s['type'] == 'definition-ref':
-            if next_s['schema_ref'] in visited:
+        visited = set()
+        while 'schema_ref' in next_s and next_s['type'] == 'definition-ref':
+            next_ref = next_s['schema_ref']
+            if next_ref in visited:
                 raise PydanticUserError(
                     f'{ref} contains a circular reference to itself.', code='circular-reference-schema'
                 )
-
-            visited.add(next_s['schema_ref'])
-            ref_counts[next_s['schema_ref']] += 1
-            next_s = definitions[next_s['schema_ref']]
-
+            visited.add(next_ref)
+            ref_counts[next_ref] += 1
+            next_s = definitions[next_ref]
         recurse(next_s, count_refs)
         current_recursion_ref_count[ref] -= 1
         return s
 
     schema = walk_core_schema(schema, count_refs, copy=False)
 
-    assert all(c == 0 for c in current_recursion_ref_count.values()), 'this is a bug! please report it'
+    if any(c != 0 for c in current_recursion_ref_count.values()):
+        raise RuntimeError('This is a bug! Please report it.')
 
     def can_be_inlined(s: core_schema.DefinitionReferenceSchema, ref: str) -> bool:
-        if ref_counts[ref] > 1:
-            return False
-        if involved_in_recursion.get(ref, False):
+        if ref_counts[ref] > 1 or ref in involved_in_recursion:
             return False
-        if 'serialization' in s:
-            return False
-        if 'metadata' in s:
-            metadata = s['metadata']
-            for k in [
+        metadata = s.get('metadata', {})
+        return 'serialization' not in s and not any(
+            k in metadata
+            for k in (
                 *CoreMetadata.__annotations__.keys(),
                 'pydantic.internal.union_discriminator',
                 'pydantic.internal.tagged_union_tag',
-            ]:
-                if k in metadata:
-                    # we need to keep this as a ref
-                    return False
-        return True
+            )
+        )
 
     def inline_refs(s: core_schema.CoreSchema, recurse: Recurse) -> core_schema.CoreSchema:
-        # Assume there are no infinite loops, because we already checked for that in `count_refs`
-        while s['type'] == 'definition-ref':
+        while 'schema_ref' in s and s['type'] == 'definition-ref':
```

Comment:
```suggestion
        # Assume there are no infinite loops, because we already checked for that in `count_refs`
        while 'schema_ref' in s and s['type'] == 'definition-ref':
```